File: Tutorials/Codes/Modern_Control_Course/Chapter11/Lesson5/Chapter11_Lesson5.m
% Chapter11_Lesson5.m
% LTV controllability Gramian using MATLAB ode45.
% The same ODE can be implemented in Simulink by using a MATLAB Function
% block for A(t), B(t) and an Integrator block for the vectorized W dynamics.

clear; clc;
t0 = 0.0;
tf = 4.0;
n = 2;

Phi0 = eye(n);
W0 = zeros(n);
z0 = [Phi0(:); W0(:)];
opts = odeset('RelTol',1e-9,'AbsTol',1e-11);
[t,z] = ode45(@(t,z) augmented_ode(t,z,n), [t0 tf], z0, opts);

zf = z(end,:).';
Phi = reshape(zf(1:n*n), n, n);
Wc = reshape(zf(n*n+1:end), n, n);
Wc = 0.5*(Wc + Wc.');

disp('Phi(tf,t0) =');
disp(Phi);
disp('Wc(t0,tf) =');
disp(Wc);
disp('eig(Wc) =');
disp(eig(Wc));

if min(eig(Wc)) > 1e-8
    disp('The LTV system is controllable on this interval.');
else
    disp('The test is singular or ill-conditioned on this interval.');
end

function dz = augmented_ode(t,z,n)
    Phi = reshape(z(1:n*n), n, n);
    W = reshape(z(n*n+1:end), n, n);
    At = A_ltv(t);
    Bt = B_ltv(t);
    dPhi = At*Phi;
    dW = At*W + W*At.' + Bt*Bt.';
    dz = [dPhi(:); dW(:)];
end

function At = A_ltv(t)
    At = [0, 1; -(2.0 + 0.60*sin(t)), -0.25];
end

function Bt = B_ltv(t)
    Bt = [0; 1.0 + 0.45*cos(t)];
end

File: Tutorials/Codes/Modern_Control_Course/Chapter11/Lesson5/Chapter11_Lesson5.py
"""
Chapter11_Lesson5.py
Time-varying controllability for a 2-state LTV system.

This script integrates the state transition matrix Phi(t,t0) and the
finite-horizon controllability Gramian Wc(t0,t) using the differential
Lyapunov equation

    dW/dt = A(t) W + W A(t)^T + B(t) B(t)^T,  W(t0)=0.

A positive definite Wc on [t0,tf] indicates complete controllability on
that interval for this continuous-time LTV system.
"""

import numpy as np
from scipy.integrate import solve_ivp


def A(t: float) -> np.ndarray:
    """Time-varying dynamics matrix."""
    return np.array([
        [0.0, 1.0],
        [-(2.0 + 0.60 * np.sin(t)), -0.25],
    ])


def B(t: float) -> np.ndarray:
    """Time-varying input matrix."""
    return np.array([
        [0.0],
        [1.0 + 0.45 * np.cos(t)],
    ])


def augmented_ode(t: float, z: np.ndarray, n: int) -> np.ndarray:
    """ODE for Phi(t,t0) and Wc(t0,t)."""
    Phi = z[: n * n].reshape(n, n)
    W = z[n * n :].reshape(n, n)
    At = A(t)
    Bt = B(t)

    dPhi = At @ Phi
    dW = At @ W + W @ At.T + Bt @ Bt.T
    return np.concatenate([dPhi.reshape(-1), dW.reshape(-1)])


def compute_gramian(t0: float, tf: float, n: int = 2):
    Phi0 = np.eye(n)
    W0 = np.zeros((n, n))
    z0 = np.concatenate([Phi0.reshape(-1), W0.reshape(-1)])

    sol = solve_ivp(
        fun=lambda t, z: augmented_ode(t, z, n),
        t_span=(t0, tf),
        y0=z0,
        rtol=1e-9,
        atol=1e-11,
        dense_output=False,
    )
    if not sol.success:
        raise RuntimeError(sol.message)

    zf = sol.y[:, -1]
    Phi_tf_t0 = zf[: n * n].reshape(n, n)
    Wc = zf[n * n :].reshape(n, n)
    Wc = 0.5 * (Wc + Wc.T)  # remove round-off asymmetry
    return Phi_tf_t0, Wc


def is_controllable(Wc: np.ndarray, tol: float = 1e-8) -> bool:
    eigvals = np.linalg.eigvalsh(Wc)
    return bool(np.min(eigvals) > tol)


if __name__ == "__main__":
    for tf in [0.25, 0.50, 1.00, 2.00, 4.00]:
        Phi, Wc = compute_gramian(0.0, tf)
        eigvals = np.linalg.eigvalsh(Wc)
        print(f"Interval [0, {tf:.2f}]")
        print("Phi(tf,t0)=\n", Phi)
        print("Wc(t0,tf)=\n", Wc)
        print("Eigenvalues of Wc:", eigvals)
        print("Controllable on this interval?", is_controllable(Wc))
        print("-" * 60)
